capsules/extra/src/pressure.rs

```diff
@@ -0,0 +1,146 @@
+//!
+//! * `Ok(())`:    The operation has been successful.
+//! * `BUSY`:      The driver is busy.
+//! * `ENOSUPPORT`: Invalid `cmd`.
+//! * `NOMEM`:     No sufficient memory available.
+//! * `INVAL`:     Invalid address of the buffer or other error.
+//!
+//! Usage
+//! -----
+//!
+//! You need a device that provides the `hil::sensors::PressureDriver` trait.
+//!
+//! ```rust
+//! # use kernel::static_init;
+//!
+//! let pressure = static_init!(
+//!        capsules::temperature::PressureSensor<'static>,
+//!        capsules::temperature::PressureSensor::new(si7021,
+//!                                                 board_kernel.create_grant(&grant_cap)));
+//!
+//! kernel::hil::sensors::PressureDriver::set_client(si7021, pressure);
+//! ```
+
+use core::cell::Cell;
+
+use kernel::grant::{AllowRoCount, AllowRwCount, Grant, UpcallCount};
+use kernel::hil;
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::{ErrorCode, ProcessId};
+
+/// Syscall driver number.
+use capsules_core::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Pressure as usize;
+
+#[derive(Default)]
+pub struct App {
+    subscribed: bool,
+}
+
+pub struct PressureSensor<'a, T: hil::sensors::PressureDriver<'a>> {
+    driver: &'a T,
+    apps: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    busy: Cell<bool>,
+}
+
+impl<'a, T: hil::sensors::PressureDriver<'a>> PressureSensor<'a, T> {
+    pub fn new(
+        driver: &'a T,
+        apps: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    ) -> PressureSensor<'a, T> {
+        PressureSensor {
+            driver: driver,
+            apps: apps,
+            busy: Cell::new(false),
+        }
+    }
+
+    fn enqueue_command(&self, processid: ProcessId) -> CommandReturn {
+        self.apps
+            .enter(processid, |app, _| {
+                app.subscribed = true;
+                if !self.busy.get() {
+                    let res = self.driver.read_atmospheric_pressure();
+                    if let Ok(err) = ErrorCode::try_from(res) {
+                        CommandReturn::failure(err)
+                    } else {
+                        self.busy.set(true);
+                        CommandReturn::success()
+                    }
+                } else {
+                    CommandReturn::success()
+                }
+            })
+            .unwrap_or_else(|err| CommandReturn::failure(err.into()))
+    }
+}
+
+impl<'a, T: hil::sensors::PressureDriver<'a>> hil::sensors::PressureClient for PressureSensor<'a, T> {
+    fn callback(&self, pressure: Result<i32, ErrorCode>) {
+        self.busy.set(false);
+        if let Ok(pressure_value) = pressure {
+            for cntr in self.apps.iter() {
+                cntr.enter(|app, upcalls| {
+                    if app.subscribed {
+                        app.subscribed = false;
+                        upcalls
+                            .schedule_upcall(0, (pressure_value as usize, 0, 0))
```

Comment:
I didn't catch this the first time, but we don't want to repeat mistakes from the temp/humidity drivers here. The first argument should be a return code instead of the actual pressure value, and the pressure value should be the second argument to the upcall.